indexer/pkg/storage/postgres.go

```diff
@@ -369,14 +376,294 @@ func (d *PostgresStorage) BatchInsertCCVData(ctx context.Context, ccvDataList []
 	}
 
 	// Increment the verification records counter by the number of rows actually inserted
-	for i := int64(0); i < rowsAffected; i++ {
+	for range rowsAffected {
 		d.monitoring.Metrics().IncrementVerificationRecordsCounter(ctx)
 	}
 
 	d.monitoring.Metrics().RecordStorageWriteDuration(ctx, time.Since(startInsertMetric))
 	return nil
 }
 
+// BatchInsertMessages implements common.IndexerStorage.
+func (d *PostgresStorage) BatchInsertMessages(ctx context.Context, messages []common.MessageWithMetadata) error {
+	if len(messages) == 0 {
+		return nil
+	}
+
+	startInsertMetric := time.Now()
+	d.mu.Lock()
+	defer d.mu.Unlock()
+
+	// Build batch insert query with multiple value sets
+	query := `
+		INSERT INTO indexer.messages (
+			message_id,
+			message,
+			status,
+			lastErr,
+			source_chain_selector,
+			dest_chain_selector,
+			ingestion_timestamp
+		) VALUES
+	`
+
+	args := make([]any, 0, len(messages)*7)
+	valueClauses := make([]string, 0, len(messages))
+
+	for i, msg := range messages {
+		// Serialize message to JSON
+		messageJSON, err := json.Marshal(msg.Message)
+		if err != nil {
+			d.monitoring.Metrics().RecordStorageInsertErrorsCounter(ctx)
+			return fmt.Errorf("failed to marshal message to JSON at index %d: %w", i, err)
+		}
+
+		// Calculate parameter positions for this row
+		baseIdx := i * 7
+		valueClause := fmt.Sprintf("($%d, $%d, $%d, $%d, $%d, $%d, $%d)",
+			baseIdx+1, baseIdx+2, baseIdx+3, baseIdx+4, baseIdx+5, baseIdx+6, baseIdx+7)
```

Comment:
This code here (and elsewhere in this file) kinda feels brittle, what with the manual index manipulation. I wonder if an actual ORM would make this less prone to potential things like SQL injection or off-by-one errors

Reply:
Even using named fields like we do in Chainlink would be an improvement IMO (e.g. instead of `$1, $2` it'd be `:first_field, :second_field`. I think `sqlx` is the library for it